diff --git a/DESCRIPTION b/DESCRIPTION
index 1823a59..a598180 100644
--- a/DESCRIPTION
+++ b/DESCRIPTION
@@ -12,7 +12,7 @@ Author: Rafael A. Irizarry <rafa@jhu.edu>, Laurent Gautier
         Christopher Workman <workman@cbs.dtu.dk>, John Zhang
 Maintainer: Rafael A. Irizarry <rafa@jhu.edu>
 Depends: R (>= 2.8.0), BiocGenerics (>= 0.1.12), Biobase (>= 2.5.5)
-Imports: affyio (>= 1.13.3), BiocInstaller, graphics, grDevices,
+Imports: affyio (>= 1.13.3), graphics, grDevices,
         methods, preprocessCore, stats, utils, zlibbioc
 Suggests: tkWidgets (>= 1.19.0), affydata, widgetTools
 LinkingTo: preprocessCore
diff --git a/NAMESPACE b/NAMESPACE
index ee0ae10..45a3245 100644
--- a/NAMESPACE
+++ b/NAMESPACE
@@ -31,8 +31,6 @@ importFrom(Biobase, addVigs2WinMenu, assayDataElementReplace,
 
 importMethodsFrom(BiocGenerics, normalize)
 
-importFrom(BiocInstaller, biocinstallRepos)
-
 importFrom(graphics, abline, axis, barplot, frame, hist,
            lines, matplot, mtext, pairs, par, plot, points,
            smoothScatter, text)
diff --git a/R/getCDFenv.R b/R/getCDFenv.R
index 2894f80..865998b 100644
--- a/R/getCDFenv.R
+++ b/R/getCDFenv.R
@@ -69,6 +69,7 @@ cdfFromBioC <- function(cdfname, lib=.libPaths()[1], verbose=TRUE) {
                   sep=""))
     if (verbose)
       print(paste("Checking to see if your internet connection works ..."))
+    stop("automatic loading of cdf packages disabled in Nix install. Use r_ecosystem_track")
     if (testBioCConnection()) {
         ## Check for file permissions
         if (file.access(lib, mode=0) < 0) {
