diff --git a/configure b/configure
index 7f7f3eb..5efeb2b 100755
--- a/configure
+++ b/configure
@@ -5,7 +5,7 @@
 # R CMD INSTALL --configure-vars='INCLUDE_DIR=/.../include LIB_DIR=/.../lib'
 
 # Library settings
-PKG_CONFIG_NAME="tesseract"
+PKG_CONFIG_NAME="tesseract lept"
 PKG_DEB_NAME="libtesseract-dev libleptonica-dev"
 PKG_RPM_NAME="tesseract-devel leptonica-devel"
 PKG_CSW_NAME="libtesseract_dev"
@@ -55,7 +55,7 @@ CXXFLAGS=`${R_HOME}/bin/R CMD config CXXFLAGS`
 CPPFLAGS=`${R_HOME}/bin/R CMD config CPPFLAGS`
 
 # Test configuration
-${CXXCPP} ${CPPFLAGS} ${PKG_CFLAGS} ${CXXFLAGS} src/test.h >/dev/null 2>&1
+${CXXCPP} ${CPPFLAGS} ${PKG_CFLAGS} ${CXXFLAGS} src/test.h 2>&1
 
 # Customize the error
 if [ $? -ne 0 ]; then
